homeassistant/components/zha/core/discovery.py

```diff
@@ -66,21 +78,86 @@
 _LOGGER = logging.getLogger(__name__)
 
 
+QUIRKS_ENTITY_META_TO_ENTITY_CLASS = {
+    (
+        Platform.BUTTON,
+        WriteAttributeButtonMetadata,
+        EntityType.CONFIG,
+    ): button.ZHAAttributeButton,
+    (Platform.BUTTON, ZCLCommandButtonMetadata, EntityType.CONFIG): button.ZHAButton,
+    (
+        Platform.BUTTON,
+        ZCLCommandButtonMetadata,
+        EntityType.DIAGNOSTIC,
+    ): button.ZHAButton,
+    (
+        Platform.BINARY_SENSOR,
+        BinarySensorMetadata,
+        EntityType.CONFIG,
+    ): binary_sensor.BinarySensor,
+    (
+        Platform.BINARY_SENSOR,
+        BinarySensorMetadata,
+        EntityType.DIAGNOSTIC,
+    ): binary_sensor.BinarySensor,
+    (
+        Platform.BINARY_SENSOR,
+        BinarySensorMetadata,
+        EntityType.STANDARD,
+    ): binary_sensor.BinarySensor,
+    (Platform.SENSOR, ZCLEnumMetadata, EntityType.DIAGNOSTIC): sensor.EnumSensor,
+    (Platform.SENSOR, ZCLEnumMetadata, EntityType.STANDARD): sensor.EnumSensor,
+    (Platform.SENSOR, ZCLSensorMetadata, EntityType.DIAGNOSTIC): sensor.Sensor,
+    (Platform.SENSOR, ZCLSensorMetadata, EntityType.STANDARD): sensor.Sensor,
+    (Platform.SELECT, ZCLEnumMetadata, EntityType.CONFIG): select.ZCLEnumSelectEntity,
+    (
+        Platform.SELECT,
+        ZCLEnumMetadata,
+        EntityType.DIAGNOSTIC,
+    ): select.ZCLEnumSelectEntity,
+    (
+        Platform.NUMBER,
+        NumberMetadata,
+        EntityType.CONFIG,
+    ): number.ZHANumberConfigurationEntity,
+    (Platform.NUMBER, NumberMetadata, EntityType.DIAGNOSTIC): number.ZhaNumber,
+    (Platform.NUMBER, NumberMetadata, EntityType.STANDARD): number.ZhaNumber,
+    (
+        Platform.SWITCH,
+        SwitchMetadata,
+        EntityType.CONFIG,
+    ): switch.ZHASwitchConfigurationEntity,
+    (Platform.SWITCH, SwitchMetadata, EntityType.STANDARD): switch.Switch,
+}
+
+
 @callback
 async def async_add_entities(
     _async_add_entities: AddEntitiesCallback,
     entities: list[
         tuple[
             type[ZhaEntity],
             tuple[str, ZHADevice, list[ClusterHandler]],
+            dict[str, Any],
         ]
     ],
     **kwargs,
 ) -> None:
     """Add entities helper."""
     if not entities:
         return
-    to_add = [ent_cls.create_entity(*args, **kwargs) for ent_cls, args in entities]
+
+    def _maybe_combine_kwargs(kwargs, entity_kwargs):
+        if entity_kwargs and kwargs:
+            return entity_kwargs.update(kwargs)
```

Comment:
Might be nice to also cover this line.

Reply:
test would be contrived. the only platform that would combine them atm is update and there is no way to express update entities from quirks yet.

Reply:
The `update` method returns `None` so I think this would actually fail. You may be able to replace both cases with this?

```suggestion
            return {**entity_kwargs, **kwargs}
```